### PR TITLE
#4: Add async predict function

### PR DESCRIPTION
**Pull request by qinwf-nuan**
## Description

This pull request adds async predict function to model class.

## Tested platforms

macOS 10.12 with Node 6

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -220,9 +221,23 @@ export function matrixFromCSR(data: Float32Array, indptr: Uint32Array,
...
+function setupQueue() {
+  return async.queue((task, callback) => {
+    task.model.predictAsync(task.mat, task.mask, task.ntree, (err, res) => {
+      task.callback(err, res);
+      task = undefined;
+      callback();
+    });
+  }, 1);
+}
+const taskQueue = setupQueue();
```
Comment (jiabinf) [inline index.ts]: you can combine two lines as self executing function, it's safer.

File: test/base.js [hunk 2]
```diff
@@ -114,4 +116,69 @@ describe('base', () => {
...
+    }
+    async.parallelLimit(tasks, 1000, (err) => {
+      console.log('async.parallel tasks end');
+      expect(err).to.be.null;
+      done();
+    });
+    console.log('async.parallel tasks begin');
+  });
+
+  it('async error', function (done) {
```
Comment (jiabinf) [inline test/base.js:177]: any other known error we can cover in testcases?

Comment (qinwf-nuan) [reply test/base.js:177]: Added restoreMatrix error test.

**Comment by jiabinf** [comment]
LGTM

**Comment by qinwf-nuan** [comment]
The cause of the Travis failure is a potential bug in xgboost library. We can reproduce this bug with python port of xgboost. 

I will try to submit an issue to the xgboost repo to see if they will fix it. 

The xgboost library should not abort the main program if the DMatrix file is not valid.

```python3
import xgboost
xgboost(''./test/data/iris.xg.model")

[[[[10:49:5010:49:5010:49:5010:49:50] ] ] ] dmlc-core/include/dmlc/./logging.hdmlc-core/include/dmlc/./logging.hdmlc-core/include/dmlc/./logging.hdmlc-core/include/dmlc/./logging.h::::300300300300: : : : [10:49:50] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1)

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000010bae8e89 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x000000010be76f80 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16
[10:49:50] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1)

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000010bae8e89 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x000000010be76f80 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16
[10:49:50] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1)

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000010bae8e89 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x000000010be76f80 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16
[10:49:50] src/data/././strtonum.h:141: Check failed: sign == true (0 vs. 1)

Stack trace returned 2 entries:
[bt] (0) 0   libxgboost.dylib                    0x000000010bae8e89 _ZN4dmlc15LogMessageFatalD1Ev + 41
[bt] (1) 1   libstdc++.6.dylib                   0x000000010be76f80 _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE + 16




terminate called recursively
terminate called recursively
[1]    79928 abort      python3
```

**Comment by qinwf-nuan** [comment]
A bug report post to https://github.com/dmlc/xgboost/issues/2849

**Pull request merged**